sql/schemachanger: Enable column type changes for virtual computed columns

spilchen opened the pull request:
This update introduces support for altering the column type of virtual computed columns in the declarative schema changer (DSC). Previously, such operations were not allowed.

A key characteristic of virtual columns is that type changes never require a backfill or column rewrite, as their values are computed dynamically during access. This change includes adjustments to handle type changes that would typically involve a rewrite as simple metadata updates instead.

Additionally, the USING expression in ALTER TABLE statements is now explicitly disallowed for virtual columns, as no data rewriting occurs. Attempts to use this option will result in an error.

Epic: CRDB-25314
Closes: #125840
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135315)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice work! just had very minor stylistic comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135315#-:-OC5BNo4A4vdn9tYQ9Aq:b-pfbn5u)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 1699 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135315#-OC5AWu8EGti0PtG4MBF:-OC5AWu8EGti0PtG4MBG:b-hfbhg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f8d67f845ade30dbcd40f89f6f516de35baa2d/pkg/sql/logictest/testdata/logic_test/alter_column_type#L1699)):*
> ```
> skipif config local-legacy-schema-changer
> skipif config local-mixed-24.1
> skipif config local-mixed-24.2
> ```

super nit: the skips can be rewritten so they're on one line now, thanks to a recent patch that was just merged.
___
*[`pkg/sql/schemachange/alter_column_type.go` line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135315#-OC5AyKH2YasaF22v999:-OC5AyKH2YasaF22v99A:btv7x60) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f8d67f845ade30dbcd40f89f6f516de35baa2d/pkg/sql/schemachange/alter_column_type.go#L278)):*
> ```Go
> 	newType *types.T,
> 	isVirtual bool,
> ) (kind ColumnConversionKind, err error) {
> ```

super nit: this would be a tiny bit more easy to follow if we distinguished the return parameter with a name of `retErr`
___
*[`pkg/sql/schemachange/alter_column_type.go` line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135315#-OC5BG_J1GbDaTqa1hE_:-OC5BG_J1GbDaTqa1hEa:b-nnnutb) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f8d67f845ade30dbcd40f89f6f516de35baa2d/pkg/sql/schemachange/alter_column_type.go#L305)):*
> ```Go
> 			oldType.SQLString(), newType.SQLString())
> 	}
> 	return
> ```

super nit: we could avoid the naked returns in this function, though since the function is pretty small, i don't feel strongly.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Enable column type changes for virtual computed columns

This update introduces support for altering the column type of virtual
computed columns in the declarative schema changer (DSC). Previously,
such operations were not allowed.

A key characteristic of virtual columns is that type changes never
require a backfill or column rewrite, as their values are computed
dynamically during access. This change includes adjustments to handle
type changes that would typically involve a rewrite as simple metadata
updates instead.

Additionally, the USING expression in ALTER TABLE statements is now
explicitly disallowed for virtual columns, as no data rewriting occurs.
Attempts to use this option will result in an error.

Epic: CRDB-25314
Closes: #125840
Release note: none

spilchen commented:
TFTR!

bors r+
